From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <nonplay@bxteam.org>
Date: Wed, 29 Oct 2025 01:31:51 +0300
Subject: [PATCH] Implement chunk system algorithm

Adds configurable chunk system algorithms (MOONRISE, C2ME, C2ME_NEW) for determining
worker thread allocation for chunk loading and generation.

diff --git a/src/main/java/ca/spottedleaf/moonrise/common/util/MoonriseCommon.java b/src/main/java/ca/spottedleaf/moonrise/common/util/MoonriseCommon.java
index ee857c23afb54763db452981e7ed1f0eb0d2d2a0..73b84c40b9f8b582548742692a6b0ffff0f319e0 100644
--- a/src/main/java/ca/spottedleaf/moonrise/common/util/MoonriseCommon.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/util/MoonriseCommon.java
@@ -44,21 +44,11 @@ public final class MoonriseCommon {
     public static final BalancedPrioritisedThreadPool.OrderedStreamGroup SERVER_GROUP = MoonriseCommon.WORKER_POOL.createOrderedStreamGroup();
 
     public static void adjustWorkerThreads(final int configWorkerThreads, final int configIoThreads) {
-        int defaultWorkerThreads = OSNuma.getNativeInstance().getTotalCores()  / 2;
-        if (defaultWorkerThreads <= 4) {
-            defaultWorkerThreads = defaultWorkerThreads <= 3 ? 1 : 2;
-        } else {
-            defaultWorkerThreads = defaultWorkerThreads / 2;
-        }
-        defaultWorkerThreads = Integer.getInteger(PlatformHooks.get().getBrand() + ".WorkerThreadCount", Integer.valueOf(defaultWorkerThreads));
-
-        int workerThreads = configWorkerThreads;
-
-        if (workerThreads <= 0) {
-            workerThreads = defaultWorkerThreads;
-        }
-
-        final int ioThreads = Math.max(1, configIoThreads);
+        // Leaf start - Implement chunk system algorithm
+        org.dreeam.leaf.chunk.ChunkSystemAlgorithm algorithm = org.dreeam.leaf.config.modules.opt.ChunkWorkerAlgorithm.algorithm;
+        int workerThreads = algorithm.evalWorkers(configWorkerThreads, configIoThreads);
+        int ioThreads = algorithm.evalIO(configWorkerThreads, configIoThreads);
+        // Leaf end - Implement chunk system algorithm
 
         // Gale start - metrics - chunk system IO threads
         chunkSystemIOThreads = ioThreads;
@@ -68,7 +58,7 @@ public final class MoonriseCommon {
         WORKER_POOL.adjustThreadCount(workerThreads);
         IO_POOL.adjustThreadCount(ioThreads);
 
-        LOGGER.info(PlatformHooks.get().getBrand() + " is using " + workerThreads + " worker threads, " + ioThreads + " I/O threads");
+        LOGGER.info("{} is using {} worker threads, {} I/O threads ({})", PlatformHooks.get().getBrand(), workerThreads, ioThreads, algorithm.name()); // Leaf - Implement chunk system algorithm
     }
 
     public static final long IO_QUEUE_HOLD_TIME = (long)(25.0e6); // 25ms
